base/bucket_test.go

```diff
@@ -231,9 +240,189 @@ func TestGetStatsVbSeqno(t *testing.T) {
+	assert.NoError(t, err, "Client certificate should be generated")
+	saveAsCertFile(t, clientCertPath, certBytes)
+}
+
+func saveAsKeyFile(t *testing.T, filename string, key *ecdsa.PrivateKey) {
+	file, err := os.Create(filename)
+	assert.NoError(t, err)
+	defer file.Close()
+	b, err := x509.MarshalECPrivateKey(key)
+	assert.NoError(t, err)
+	err = pem.Encode(file, &pem.Block{Type: "EC PRIVATE KEY", Bytes: b})
+	assert.NoError(t, err)
+}
+
+func saveAsCertFile(t *testing.T, filename string, derBytes []byte) {
+	certOut, err := os.Create(filename)
+	assert.NoError(t, err, "No error while opening cert.pem for writing")
+	err = pem.Encode(certOut, &pem.Block{Type: "CERTIFICATE", Bytes: derBytes})
+	assert.NoError(t, err, "No error while writing data to cert.pem")
+	err = certOut.Close()
+	assert.NoError(t, err, "No error while closing cert.pem")
+}
+
+func deletePaths(paths []string) {
+	for _, path := range paths {
+		os.Remove(path)
+		log.Printf("Deleted: %v", path)
+	}
+}
+
+func TestTLSConfig(t *testing.T) {
+	// Mock fake root CA and client certificates for verification
+	mockCertificatesAndKeys(t)
+
+	// Simulate error creating tlsConfig for DCP processing
+	spec := BucketSpec{
+		Server:     "http://localhost:8091",
+		Certpath:   "/var/lib/couchbase/unknown.client.cert",
+		Keypath:    "/var/lib/couchbase/unknown.client.key",
+		CACertPath: "/var/lib/couchbase/unknown.root.ca.pem",
+	}
+	conf := spec.TLSConfig()
+	assert.Nil(t, conf)
+
+	// Simulate valid configuration scenario with fake mocked certificates and keys;
+	spec = BucketSpec{Certpath: clientCertPath, Keypath: clientKeyPath, CACertPath: rootCertPath}
+	conf = spec.TLSConfig()
+	assert.NotEmpty(t, conf)
+	assert.NotNil(t, conf.RootCAs)
+	assert.Equal(t, 1, len(conf.Certificates))
+	assert.False(t, conf.InsecureSkipVerify)
+
+	// Check TLSConfig with no CA certificate; InsecureSkipVerify should be true
+	spec = BucketSpec{Certpath: clientCertPath, Keypath: clientKeyPath}
+	conf = spec.TLSConfig()
+	assert.NotEmpty(t, conf)
+	assert.True(t, conf.InsecureSkipVerify)
+	assert.Equal(t, 1, len(conf.Certificates))
+	assert.Nil(t, conf.RootCAs)
+
+	// Check TLSConfig by providing invalid root CA certificate; provide root certificate key path
+	// instead of root CA certificate. It should throw "can't append certs from PEM" error.
+	spec = BucketSpec{Certpath: clientCertPath, Keypath: clientKeyPath, CACertPath: rootKeyPath}
+	conf = spec.TLSConfig()
+	assert.Empty(t, conf)
+
+	// Provide invalid client certificate key along with valid certificate; It should fail while
+	// trying to add key and certificate to config as x509 key pair;
+	spec = BucketSpec{Certpath: clientCertPath, Keypath: rootKeyPath, CACertPath: rootCertPath}
+	conf = spec.TLSConfig()
+	assert.Empty(t, conf)
+
+	// Remove the keys and certificates after verification
+	deletePaths([]string{rootKeyPath, rootCertPath, clientKeyPath, clientCertPath})
```

Comment:
Should this be in a defer?  If the test fails, in most cases I'd expect we'd want to clean up os.TempDir() for any subsequent run.

Reply:
Yeah, this would be a cleaver idea; thank you Adam.

Reply:
Changed are committed to the latest revision.